src/compose/service-manager.ts

```diff
@@ -0,0 +1,662 @@
+		await this.start(target);
+		await this.waitToKill(
+			current,
+			target.config.labels['io.balena.update.handover-timeout'],
+		);
+		await this.kill(current);
+	}
+
+	public async killAllLegacy(): Promise<void> {
+		// Containers haven't been normalized (this is an updated supervisor)
+		// so we need to stop and remove them
+		const supervisorImageId = (await this.docker
+			.getImage(constants.supervisorImage)
+			.inspect()).Id;
+
+		for (const container of await this.docker.listContainers({ all: true })) {
+			if (container.ImageID !== supervisorImageId) {
+				await this.killContainer(container.Id, {
+					serviceName: 'legacy',
+				});
+			}
+		}
+	}
+
+	public kill(service: Service, opts: KillOpts = {}) {
+		if (service.containerId == null) {
+			throw new InternalInconsistencyError(
+				`Attempt to kill container without containerId! Service :${service}`,
+			);
+		}
+		return this.killContainer(service.containerId, service, opts);
+	}
+
+	public async remove(service: Service) {
+		this.logger.logSystemEvent(LogTypes.removeDeadService, { service });
+		const existingService = await this.get(service);
+
+		if (existingService.containerId == null) {
+			throw new InternalInconsistencyError(
+				`No containerId provided for service ${
+					service.serviceName
+				} in ServiceManager.updateMetadata. Service: ${service}`,
+			);
+		}
+
+		try {
+			await this.docker
+				.getContainer(existingService.containerId)
+				.remove({ v: true });
+		} catch (e) {
+			if (NotFoundError(e)) {
+				return;
+			} else {
+				this.logger.logSystemEvent(LogTypes.removeDeadServiceError, {
+					service,
+					error: e,
+				});
+				throw e;
+			}
+		}
+	}
+
+	public async create(service: Service) {
+		const mockContainerId = this.config.newUniqueKey();
+		try {
+			const existing = await this.get(service);
+			if (existing.containerId == null) {
+				throw new InternalInconsistencyError(
+					`No containerId provided for service ${
+						service.serviceName
+					} in ServiceManager.updateMetadata. Service: ${service}`,
+				);
+			}
+			return this.docker.getContainer(existing.containerId);
+		} catch (e) {
+			if (!NotFoundError(e)) {
+				this.logger.logSystemEvent(LogTypes.installServiceError, {
+					service,
+					error: e,
+				});
+				throw e;
+			}
+
+			const deviceName = await this.config.get('name');
+			if (!isValidDeviceName(deviceName)) {
+				throw new Error(
+					'The device name contains a newline, which is unsupported by balena. ' +
+						'Please fix the device name',
+				);
+			}
+
+			const conf = service.toDockerContainer({ deviceName });
+			const nets = service.extraNetworksToJoin();
+
+			this.logger.logSystemEvent(LogTypes.installService, { service });
+			this.reportNewStatus(mockContainerId, service, 'Installing');
+
+			const container = await this.docker.createContainer(conf);
+			service.containerId = container.id;
+
+			await Promise.all(
+				_.map(nets, (endpointConfig, name) =>
```

Comment:
I assume this is an object hence the `_.map`?

Reply:
It is, coming from https://github.com/balena-io/balena-supervisor/blob/master/src/compose/service.ts#L690